File: open_spiel/public_states/algorithms/cfr_test.cc
// Copyright 2019 DeepMind Technologies Ltd. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "open_spiel/public_states/algorithms/cfr.h"

#include <cmath>
#include <iostream>

#include "open_spiel/public_states/public_states.h"
#include "open_spiel/algorithms/expected_returns.h"
#include "open_spiel/algorithms/history_tree.h"
#include "open_spiel/algorithms/tabular_exploitability.h"
#include "open_spiel/public_states/games/kuhn_poker.h"
#include "open_spiel/games/kuhn_poker.h"
#include "open_spiel/spiel.h"
#include "open_spiel/spiel_utils.h"

namespace open_spiel {
namespace public_states {
namespace algorithms {
namespace {

void CheckNashKuhnPoker(const Game& game, const Policy& policy) {
  const std::vector<double> game_value =
      ::open_spiel::algorithms::ExpectedReturns(
          *game.NewInitialState(), policy, /*depth_limit=*/-1,
          /*use_infostate_get_policy=*/false);

  // 1/18 is the Nash value. See https://en.wikipedia.org/wiki/Kuhn_poker
  constexpr double nash_value = 1.0 / 18.0;
  constexpr double eps = 1e-3;

  SPIEL_CHECK_EQ(2, game_value.size());
  SPIEL_CHECK_FLOAT_NEAR(game_value[0], -nash_value, eps);
  SPIEL_CHECK_FLOAT_NEAR(game_value[1], nash_value, eps);
}

void CheckExploitabilityKuhnPoker(const Game& game, const Policy& policy) {
  SPIEL_CHECK_LE(::open_spiel::algorithms::Exploitability(game, policy), 0.05);
}

void CFRTest_KuhnPoker() {
  std::shared_ptr<const GameWithPublicStates> game =
      LoadGameWithPublicStates("kuhn_poker");
  CFRPublicStatesSolver solver(*game);
  for (int i = 0; i < 300; i++) {
    solver.RunIteration();
  }

  const std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
  CheckNashKuhnPoker(*game->GetBaseGame(), *average_policy);
  CheckExploitabilityKuhnPoker(*game->GetBaseGame(), *average_policy);
}


void CFRPlusTest_KuhnPoker() {
  std::shared_ptr<const GameWithPublicStates> game =
      LoadGameWithPublicStates("kuhn_poker");
  CFRPlusPublicStatesSolver solver(*game);
  for (int i = 0; i < 200; i++) {
    solver.RunIteration();
  }
  const std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
  CheckNashKuhnPoker(*game->GetBaseGame(), *average_policy);
  CheckExploitabilityKuhnPoker(*game->GetBaseGame(), *average_policy);
}

// TODO(author13): Implement terminal values for multi-player kuhn and enable
//              add multi-player CFR tests.


}  // namespace
}  // namespace algorithms
}  // namespace public_states
}  // namespace open_spiel

namespace algorithms = open_spiel::public_states::algorithms;

int main(int argc, char** argv) {
  algorithms::CFRTest_KuhnPoker();
  algorithms::CFRPlusTest_KuhnPoker();
}

File: open_spiel/open_spiel/python/examples/meta_cfr/sequential_games/cfr_test.py
# Copyright 2022 DeepMind Technologies Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests counterfactual regret minimization."""

from absl.testing import absltest
from absl.testing import parameterized

from open_spiel.python.examples.meta_cfr.sequential_games import cfr
from open_spiel.python.examples.meta_cfr.sequential_games import game_tree_utils as trees
from open_spiel.python.examples.meta_cfr.sequential_games import openspiel_api


def _uniform_policy(size):
  if size > 0:
    return [1./size]*size
  return []


class CfrTest(parameterized.TestCase):

  @parameterized.named_parameters(('kuhn_poker_test', 'kuhn_poker'),
                                  ('leduc_poker_test', 'leduc_poker'))
  def test_zero_policy_is_uniform(self, game):
    config = {'players': 2}
    cfr_game_tree = trees.build_game_tree(
        openspiel_api.WorldState(
            game_name=game, config=config, perturbation=False))
    cfr.compute_cfr_values(cfr_game_tree, 1)
    infostates_p1 = list(cfr_game_tree.all_infostates_map[1].values())
    infostates_p2 = list(cfr_game_tree.all_infostates_map[2].values())
    with self.subTest('player_1_initial_policy'):
      for i in range(len(infostates_p1)):
        self.assertListEqual(
            list(infostates_p1[i].policy.values()),
            _uniform_policy(len(infostates_p1[i].policy.values())))
    with self.subTest('player_2_initial_policy'):
      for i in range(len(infostates_p2)):
        self.assertListEqual(
            list(infostates_p2[i].policy.values()),
            _uniform_policy(len(infostates_p2[i].policy.values())))

  def test_cfr_leduc_poker(self):
    config = {'players': 2}
    exploitability_error = 0.2
    cfr_game_tree = trees.build_game_tree(
        openspiel_api.WorldState(
            game_name='leduc_poker', config=config, perturbation=False))
    best_response_value_p1, best_response_value_p2 = cfr.compute_cfr_values(
        cfr_game_tree, 20)
    last_best_response_value_player_1 = best_response_value_p1[-1]
    last_best_response_value_player_2 = best_response_value_p2[-1]
    exploitability = (last_best_response_value_player_1 +
                      last_best_response_value_player_2) / 2
    # Exploitability values are computed using OpenSpiel cfr
    self.assertLessEqual(exploitability, 0.59 + exploitability_error)

  def test_cfr_kuhn_poker(self):
    config = {'players': 2}
    exploitability_error = 0.2
    cfr_game_tree = trees.build_game_tree(
        openspiel_api.WorldState(
            game_name='kuhn_poker', config=config, perturbation=False))
    best_response_value_p1, best_response_value_p2 = cfr.compute_cfr_values(
        cfr_game_tree, 20)
    last_best_response_value_player_1 = best_response_value_p1[-1]
    last_best_response_value_player_2 = best_response_value_p2[-1]
    exploitability = (last_best_response_value_player_1 +
                      last_best_response_value_player_2) / 2
    # Exploitability values are computed using OpenSpiel cfr
    self.assertLessEqual(exploitability, 0.06 + exploitability_error)


if __name__ == '__main__':
  absltest.main()
